### ISSUE TITLE
[Format][Docs] Describe C device interface on C data interface and C stream interface docs pages

### ISSUE THREAD

**Issue by ianmcook** [apache/arrow#44535]
### Describe the enhancement requested

The following facts are not immediately obvious to readers of the C data interface and C stream interface docs pages:
- The C data interface is a CPU-only subset of the C device data interface.
- The C stream interface is a CPU-only subset of the C device stream interface.
- We will never remove the existing CPU-only C data/stream interfaces (because we must keep them to preserve compatibility with existing systems)...
- ... but when we add new structure definitions (for example as in #43632), we will add them _only_ to the C _device_ data/stream interfaces (to reduce duplication of efforts).
- In other words, the CPU-only C data/stream interfaces are in "maintenance mode."
- This will force developers that are currently using the CPU-only C data/stream interfaces and who wish to use the new structures to switch over to using the C _device_ data/stream interfaces. This is simple enough to do.

To make this clearer, we should add a note box to to `CDataInterface.rst` and `CStreamInterface.rst` to explain some of this and to link to `CDeviceDataInterface.rst`. They should clearly state that the CPU-only C data/stream interfaces are in "maintenance mode."

### Component(s)

Documentation, Format

**Comment by jorisvandenbossche** [comment]
>  The note boxes should clearly state that the CPU-only C data/stream interfaces are in "maintenance mode."

I think we should be careful about the wording, so it is clear that almost all content on those pages are still relevant, because even when implementing the device interface, a lot of the content you need is actually explained on the C data interface / C stream interface pages.

Somewhat related to that, I have thought in the past if we should consolidate the C data interface / C stream interface pages. I know historically they have been added (and voted) as two separate interface. But I think keeping them separate actually is also confusing (because again many of the details for the stream interface are then actually explained on the data interface page)

**Comment by ianmcook** [comment]
> we should be careful about the wording
Agreed, thanks @jorisvandenbossche 

> ... we should consolidate the C data interface / C stream interface pages ... keeping them separate actually is also confusing

Also agreed. The fact that the C device interface extends both the C data and C stream interfaces—but the former is documented on one page while the latter are documented on two—seems quite confusing.
